File: 1390-four-divisors/1390-four-divisors.java
class Solution {
    private static HashMap<Integer,Set<Integer>> map = new HashMap<>();
    static {
        for (int i = 1; i <= 100000; i++) {
            Set<Integer> set = new HashSet<>();
            for (int j = 1; j * j <= i; j++) {
                if (i % j == 0) {
                    set.add(j);
                    set.add(i/j);
                    if (set.size() > 4) break;
                }
            }
            if (set.size() == 4) {
                map.put(i,set);
            }
        }
    }

    public int sumFourDivisors(int[] nums) {
        int res = 0;
        for (int x: nums) {
            if (map.containsKey(x)) {
                for (int y: map.get(x)) {
                    res += y;
                }
            }
        }
        return res;
    }
}

File: 1390-four-divisors/1390-four-divisors.py
yes = {}
for i in range(1,100001):
    c = set()
    for j in range(1, int(i ** 0.5) + 1):
        if i % j == 0:
            if j * j == i:
                c.add(j)
            else:
                c.add(j)
                c.add(i // j)
        if len(c) > 4:
            break
    if len(c) == 4:
        yes[i] = c
        
class Solution:
    def sumFourDivisors(self, nums: List[int]) -> int:
        res = 0
        for i in nums:
            if i in yes:
                res += sum(yes[i])
        return res